Fix #422: Markdown formatting for chat history

jeanlucthumm opened the pull request:
![20240114_07h42m31s_grim](https://github.com/TheR1D/shell_gpt/assets/4934853/908908e6-afca-475e-a9fe-efd621dc8a7e)

jeanlucthumm commented:
@TheR1D is there any procedure I need to do for assigning a reviewer?

TheR1D reviewed (changes requested):
Thanks @jeanlucthumm 
I think we need to apply Markdown rendering only for "default" role conversations. Since we can have `--code` conversations, and we are using pure text ouput there, markdown will skip `\n` and print multiple lines of code in one line.
```shell
sgpt --repl --code c1
# <- print hello and print world using Python
# -> print('hello')
# -> print('world')
# <- exit()
```
```shell
sgpt --show-chat c1
...
user: print hello and print world using Python                                                                                                   
assistant: print('hello') print('world') # <---
```

jeanlucthumm commented:
Ah ok good catch. Before enabling it only for `default`, what are your thoughts on wrapping assistant messages with markdown code block if we are in code mode? I think that would look very nice and it makes sense because the assistant is outputting code after all.

We can either change how the message is stored, or we can do this only for `show_messages()` by detecting mode and whether it's an assistant message.

TheR1D commented:
> Ah ok good catch. Before enabling it only for `default`, what are your thoughts on wrapping assistant messages with markdown code block if we are in code mode? I think that would look very nice and it makes sense because the assistant is outputting code after all.
> 
> We can do this only for `show_messages()` by detecting mode and whether it's an assistant message.
> 
> The main concern here would be not outputting markdown wrapper when user is expected to redirect to file. But I think the use case for that is only this:
> 
> ```
> sgpt --code --chat c1 "a python file that says hello" > hello.py
> sgpt --code "a python file that says hello" > hello.py
> ```
> 
> Both of which do not use `show_messages`

You're right, we can use markdown code block when showing `--code` conversations with `show_messages`. May be we can apply md formatting for `--shell` conversations as well. For instance wrap assistant responses with "inline code" using backticks ``, e.g. `ls -la | sort`

jeanlucthumm commented:
Sounds good. I'll need to change how conversations are stored for this because we don't currently store the role along side the messages, only a json list of messages, so `show-chat` has no way of knowing.

My idea is to change the format from a list of messages to an object like this:

```
{
  "role": "...",
  "messages": [
    ...
  ],
}
```

This is also good for future if we ever want to store more metadata about the conversations.

Will take a look this weekend.

jeanlucthumm commented:
Ok, @TheR1D please take a look.

Some demos:
![code](https://github.com/TheR1D/shell_gpt/assets/4934853/2cf0054a-3422-43e8-ae05-eaf917db2d9d)
![default_role](https://github.com/TheR1D/shell_gpt/assets/4934853/2199aed4-b849-4b7c-9121-b65d83ae9ad8)
![shell](https://github.com/TheR1D/shell_gpt/assets/4934853/fbb95455-9ee9-41c1-84a8-f3af5b959796)

TheR1D commented:
Thank you, @jeanlucthumm. I will review it during this week. I believe we may need to make some changes.

TheR1D commented:
Hey @jeanlucthumm, thank you for the PR. After reviewing the changes, I believe this feature could be implemented more concisely. We can reuse the `initial_message` method available in `ChatHandler`. Here's an example:
```python
@classmethod
def initial_message(cls, chat_id: str) -> str:
    chat_history = cls.chat_session.get_messages(chat_id)
    return chat_history[0] if chat_history else ""
```
Then we can slightly change logic in `show_messages`.
```python
@classmethod
def show_messages(cls, chat_id: str) -> None:
    # Prints all messages from a specified chat ID to the console.
    init_message = cls.initial_message(chat_id)

    if "use and apply markdown" in init_message.lower():
        for message in cls.chat_session.get_messages(chat_id):
            if message.startswith("assistant"):
                Console().print(Markdown(message))
            else:
                typer.secho(message, fg=cfg.get("DEFAULT_COLOR"))
            typer.echo()
        return

    for index, message in enumerate(messages):
        color = "magenta" if index % 2 == 0 else "green"
        typer.secho(message, fg=color)
```
I think this is more suitable solution since we are following the system prompt, and it would be compatible with custom roles. It will not wrap `--code` conversations into a code block, but after giving it some more thought, I don't see many benefits in rendering the code using a markdown block. I think `--code` will mostly be used when a user wants to redirect clean code output elsewhere. If Markdown is required, we can prompt code generation using the default ShellGPT role. Could you also include tests to make sure it works as expected?

Thank you for interest and effort.

jeanlucthumm commented:
Sure, let's do it this way. Will take a look this weekend!

jeanlucthumm commented:
@TheR1D Ok I implemented your suggested change and also added unit tests. I also updated the documentation for custom roles so users are aware of this functionality.

<img width="1069" alt="Screenshot 2024-02-09 at 12 53 30 PM" src="https://github.com/TheR1D/shell_gpt/assets/4934853/ce1c8b00-5832-4c71-87be-c04e438c1870">
<img width="1069" alt="Screenshot 2024-02-09 at 12 54 46 PM" src="https://github.com/TheR1D/shell_gpt/assets/4934853/ffd455b0-0932-4b7f-b324-1aec037a044c">

TheR1D commented on sgpt/handlers/handler.py:
```diff
@@ -24,7 +24,7 @@ def __init__(self, role: SystemRole) -> None:
     @property
     def printer(self) -> Printer:
-        use_markdown = "APPLY MARKDOWN" in self.role.role
+        use_markdown = "apply markdown" in self.role.role.lower()
```
It was intentionally written in uppercase as `"APPLY MARKDOWN"` because we might encounter descriptions such as `"do not apply markdown"` In such cases, we will use markdown formatting for a role that, in fact, does not require markdown.

jeanlucthumm replied:
Done.

TheR1D reviewed (changes requested):
Just one minor comment, other than that, LGTM.

jeanlucthumm commented:
Ok finished!

TheR1D commented:
Thank you @jeanlucthumm, merged.
